### DISCUSSION TITLE
Missing "dcd_xxx" functions

### DISCUSSION THREAD

**Discussion by caipifrosch** [hathach/tinyusb#721] [Q&A] [answered]
Implementing tinyUsb is giving me some trouble. Maybe I am missing something here, but my MCUXpresso cannot find some functions related to usbd sources in device folder: ERROR: undefined reference to `dcd_edpt_stall'

There are two files with calls of missing dcd functions: usbd.c and usbd_control.c

Any hint where to find or include these functions?

Thanks a lot for your help in advance!

**Comment by hathach** [comment]
First of all, could you tell me your 
- setup, OS, IDE, Board & MCU
- what are you trying to do, e.g building a specific stock example or custom app, best to start with stock one
- what you have done so far

check out the makefile for example to see how the CFLAGS and files are included.

**Comment by caipifrosch** [comment, marked as answer]
Thanks a lot for your reply. I did not use any make files.

Please find detailed project information and procedure below.

1. Win10. MCUXpresso 11.3
2. tinyUsb test was using existing LPC1788 hardware
3. first I ensured that existing freeRTOS firmware project compiled and debugged ok
4. then I extracted tinyUsb Github folder to my project and added src to IDE incl. folders
5. additionally, I extracted dual LUN msc example to my project folder (choosen to be starting point)
6. example main() was changed to a init function, called from original projects initialization
7. Two defines were added to tusb_config.h:  #define CFG_TUSB_OS	OPT_OS_FREERTOS and #define CFG_TUSB_MCU 	OPT_MCU_LPC177X_8X
8. Blink timer stuff was removed
9. MCUXpresso IDE compilation fails on missing dcd_xxx functions

**Reply by hathach** [reply]
ah, it is probably my fault, since I don't have lpc177x_8x to test with, I forgot to enable it properly. can you add it into this file 
https://github.com/hathach/tinyusb/blob/master/src/portable/nxp/lpc17_40/dcd_lpc17_40.c#L29 to make it

```
#if TUSB_OPT_DEVICE_ENABLED && (CFG_TUSB_MCU == OPT_MCU_LPC175X_6X || CFG_TUSB_MCU == OPT_MCU_LPC177X_8X || CFG_TUSB_MCU == OPT_MCU_LPC40XX)
```

Note: since I haven't run this on actualy 1788 hardware, it may or may not work, please just post any issue you encounter here.  I will try to help if I could.

**Reply by hathach** [reply]
just added PR to enable this https://github.com/hathach/tinyusb/pull/722 would you mind checking it out

**Comment by caipifrosch** [comment]
OK, thanks for your quick response. Project compiles after removing second LUN msc (SRAM overflow), but windows claims "unknown usb device. error on resetting port". LC Display of my custom board is without content. Maybe something went wrong while removing second LUN, but I suspect something wrong in low level prozessor setup now (i.e. USB clocking).

**Reply by hathach** [reply]
1788 is similar to 4088, maybe this can help you out https://github.com/hathach/tinyusb/blob/master/hw/bsp/ea4088qs/ea4088qs.c

also be aware of the sram region. I remembered these are picky where the sram region. It should be defined to appropriate section by this https://github.com/hathach/tinyusb/blob/8775e89a4f7a8859d5c8bdd157e8cb10c9ed74fb/hw/bsp/ea4088qs/board.mk#L13

**Reply by caipifrosch** [reply]
Ok, even after sorting out some issues, I am still not there...

tinyUsb SRAM was moved to RAM Section 2
Hard faults were resolved after powering the LPC1788 USB module
Now tinyUsb returned from initialization and LCD showed content again
windows tries to get description information after correctly setting USB hardware pullup by mcu 
Now LCD is again without content...
-> Debugger periodically ends up in USB_IRQHandler

Seems fine to me, but I am currently not sure how to link this to tinyUsb

Any suggestion how to install the IRQ handler in tinyUsb?

**Reply by hathach** [reply]
It is in ea4088qs.c board above, please check it out.

**Reply by caipifrosch** [reply]
Okay, that is it. Got it running now.

Thanks a lot for your help!
